### ISSUE TITLE
Procedure pointers to internal procedures (part 2)

### ISSUE THREAD

**Issue by RichBarton-Arm** [flang-compiler/flang#872]
This is a new issue similar to issue [#577 ](https://github.com/flang-compiler/flang/issues/577). The example below is adapted from that issue but with the procedure taking an array of unknown length as its argument rather than an integer.
```
module m
  implicit none
  abstract interface
    subroutine int_printer(i)
      integer, dimension(:) :: i  ! Change from previous example
    end subroutine
  end interface
  procedure(int_printer), pointer :: iprinter
contains
  subroutine work
    implicit none
    integer, dimension(2) :: i1 = (/ 1, 2 /)        ! Change from previous example
    call iprinter(i1)
  end subroutine
end module
subroutine sub
  use m
  implicit none
  integer num_prints
  num_prints = 0
  iprinter => myprint
  call work
  print *, 'num_prints=', num_prints
contains
  subroutine myprint(i)
    integer, dimension(:) :: i       ! Change from previous example
    print *, i
    num_prints = num_prints + 1
  end subroutine
end subroutine

use m
implicit none
call sub
! call work  ! error
end
```

When compiled at any optimisation level with flang built from the [latest commit](https://github.com/flang-compiler/flang/commit/cb42a171b6462b2f3acd3e1841a56a1ae08291ef) on AArch64 , the resulting image segfaults:
`> ./a.out 
            1            2
Segmentation fault
`
gfortran and F18 compile this correctly.

Looking at the generated image (built with -O3) the call through iprinter in work calls to the wrong address 

> objdump -rd example.o
0000000000000004 <m_work_>:
<snip>
5c:   90000008        adrp    x8, 60 <m_work_+0x5c>
                              5c: R_AARCH64_ADR_PREL_PG_HI21  _m_0_
60:   91000108        add     x8, x8, #0x0
                              60: R_AARCH64_ADD_ABS_LO12_NC   _m_0_
64:   f9402d02        ldr     x2, [x8, #88]
68:   f9400108        ldr     x8, [x8]

This regression was caused by the commit [82683130](https://github.com/flang-compiler/flang/commit/82683130aeebdbe6d5c70044cd79ef22e54f4f22) which was part of the fix for issue #577 .

**Comment by RichBarton-Arm** [comment]
Reverting [82683130](https://github.com/flang-compiler/flang/commit/82683130aeebdbe6d5c70044cd79ef22e54f4f22) makes this go away while still fixing the original bug in #577.

**Comment by mleair** [comment]
The change in 82683130  needs a guard, for example:

if (!ABSTRACTG(sym)) {
  INTERFACEP(sym, 1);
 }

This has to do with interfaces in procedure pointers. We do not want to set the INTERFACE flag
if the symbol 	is an abstract interface. That's because the symbol name is not the same
as an actual procedure symbol name. So, we need to call the llvm_check_retval_inargs() routine in this case.

**Comment by RichBarton-Arm** [comment]
Thanks @mleair for the fix - I agree that this fixes the above example.

However, if I modify the above example to remove the abstract keyword from the interface on line 3, flang will still generate an image which segfaults when run. Here is a slightly modified example showing this.

```
module m
  implicit none
  interface
    subroutine int_printer(i)
      integer, dimension(:), intent(in) :: i
    end subroutine
  end interface
  procedure(int_printer), pointer :: iprinter
contains
  subroutine work
    implicit none
    integer, dimension(:), allocatable :: i1
    allocate(i1(2))
    call iprinter(i1)
  end subroutine
end module
subroutine sub
  use m
  implicit none
  integer num_prints
  num_prints = 0
  iprinter => myprint
  call work
  print *, 'num_prints=', num_prints
contains
  subroutine myprint(i)
    integer, dimension(:), intent(in) :: i
    print *, i
    num_prints = num_prints + 1
  end subroutine
end subroutine

use m
implicit none
call sub
! call work  ! error
end
```

gfortran accepts this just fine (although I needed to make i1 allocatable in work otherwise it rejects the code.

I'm on shaky ground here, but I think strictly speaking that int_printer is an abstract interface, i.e. it is being used as such and so perhaps the code should be rejected as the abstract keyword is not present. That said, flang (and gfortran) are happy to accept the code and treat int_printer as an abstract interface all the same. So if we are happy the code is legal enough then the segfault is the bug.

**Comment by mleair** [comment]
We are testing a fix that addresses all the cases mentioned in this issue.

**Comment by RichBarton-Arm** [comment]
@mleair do you have any update on the fix you mentioned back in February?

**Comment by mleair** [comment]
The fix just needs to be integrated.  @gklimowicz any update on the timeline for its integration?

**Comment by pawosm-arm** [comment]
Fixed by PR #1252
